lib/fluent/plugin/in_watch_process.rb

```diff
@@ -44,53 +42,237 @@ def shutdown
-          lstart = Time.parse($~[:lstart])
-          result = result.sub($~[:lstart], '')
-          keys_size -= 1
-        end
-        values = [lstart.to_s, result.chomp.strip.split(/\s+/, keys_size)]
-        data = Hash[@keys.zip(values.reject(&:empty?).flatten)]
-        data['elapsed_time'] = (Time.now - Time.parse(data['start_time'])).to_i if data['start_time']
+      @watcher.process do |data|
         next unless @lookup_user.nil? || @lookup_user.include?(data['user'])
         emit_tag = tag.dup
         filter_record(emit_tag, Fluent::Engine.now, data)
         router.emit(emit_tag, Fluent::Engine.now, data)
       end
-      io.close
     rescue StandardError => e
       log.error "watch_process: error has occured. #{e.message}"
     end
 
-    def get_ps_command
-      if OS.linux?
-        "LANG=en_US.UTF-8 && ps -ewwo lstart,user:20,pid,ppid,time,%cpu,%mem,rss,sz,s,comm,cmd"
-      elsif OS.mac?
-        "LANG=en_US.UTF-8 && ps -ewwo lstart,user,pid,ppid,time,%cpu,%mem,rss,vsz,state,comm,command"
+    module Watcher
+      def self.get_watcher(keys, command)
+        if OS.linux?
+          Linux.new(keys, command)
+        elsif OS.mac?
+          Mac.new(keys, command)
+        elsif OS.windows?
+          Windows.new(keys, command)
+        else
+          Linux.new(keys, command)
+        end
       end
-    end
 
-    module OS
-      # ref. http://stackoverflow.com/questions/170956/how-can-i-find-which-operating-system-my-ruby-program-is-running-on
-      def OS.windows?
-        (/cygwin|mswin|mingw|bccwin|wince|emx/ =~ RUBY_PLATFORM) != nil
+      class Base
+        attr_reader :keys
+        attr_reader :command
+
+        def initialize(keys, command)
+          @keys = keys
+          @command = command
+        end
+
+        def default_types
+          raise NotImplementedError, "Need to override #{self.class}##{__method__}."
+        end
+
+        def parse_line(line)
+          raise NotImplementedError, "Need to override #{self.class}##{__method__}."
+        end
+
+        def process
+          io = open_and_get_io
+          begin
+            while result = io.gets
+              yield parse_line(result)
+            end
+          ensure
+            close_io(io)
+          end
+        end
+
+        def open_and_get_io
+          io = IO.popen(@command, 'r')
+          io.gets
+          io
+        end
+
+        def close_io(io)
+          io.close
+        end
       end
+      private_constant :Base
+
+      class BaseUnixLike < Base
+        DEFAULT_KEYS = %w(start_time user pid parent_pid cpu_time cpu_percent memory_percent mem_rss mem_size state proc_name command)
+
+        def initialize(keys, command)
+          super(keys, command)
+          @keys = @keys || DEFAULT_KEYS
+        end
+
+        def default_types
+          "pid:integer,parent_pid:integer,cpu_percent:float,memory_percent:float,mem_rss:integer,mem_size:integer"
+        end
```

Comment:
It may be a personal opinion, it may be readable to use `%w()` style.

Before

```ruby
def default_types
  "pid:integer,parent_pid:integer,cpu_percent:float,   ..."
end
```

After:

```ruby
def default_types
  %w(
    pid:integer
    parent_pid:integer
    cpu_percent:float
    ...
  ).join(",")
end
```

The important point is NAME:TYPE pair is distinguishable easily.

Reply:
Thank you for the advice.
I also think it is more readable. I will fix it.

Reply:
I have fixed this in https://github.com/daipom/fluent-plugin-watch-process/pull/2/commits/e0b318de2dff6c286c1e4e43eefcb39eae77ac84.